src/brsTypes/components/RoSGNode.ts

```diff
@@ -1409,18 +1415,90 @@ export class RoSGNode extends BrsComponent implements BrsValue, BrsIterable {
+     * a node without a parent (root node). Returns the parent chain starting with root-most parent.
+     * @param node The leaf node to create the tree with
+     */
+    private createPath(node: RoSGNode): RoSGNode[] {
+        let path: RoSGNode[] = [node];
+
+        while (node.parent instanceof RoSGNode) {
+            path.push(node.parent);
+            node = node.parent;
+        }
+
+        return path.reverse();
+    }
+
     /**
      *  If on is set to true, sets the current remote control focus to the subject node,
      *  also automatically removing it from the node on which it was previously set.
-     *  If on is set to false, removes focus from the subject node if it had it
+     *  If on is set to false, removes focus from the subject node if it had it.
+     *
+     *  It also runs through all of the ancestors of the node that was focused prior to this call,
+     *  and the newly focused node, and sets the `focusedChild` field of each to reflect the new state.
      */
     private setfocus = new Callable("setfocus", {
         signature: {
             args: [new StdlibArgument("on", ValueKind.Boolean)],
             returns: ValueKind.Boolean,
         },
         impl: (interpreter: Interpreter, on: BrsBoolean) => {
-            interpreter.environment.setFocusedNode(on.toBoolean() ? this : BrsInvalid.Instance);
+            let focusedChildString = new BrsString("focusedchild");
+            let currFocusedNode = interpreter.environment.getFocusedNode();
+
+            if (on.toBoolean()) {
+                interpreter.environment.setFocusedNode(this);
+
+                // Get the focus chain, with lowest ancestor first.
+                let newFocusChain = this.createPath(this);
+
+                // If there's already a focused node somewhere, we need to remove focus
+                // from it and its ancestors.
+                if (currFocusedNode instanceof RoSGNode) {
+                    // Get the focus chain, with root-most ancestor first.
+                    let currFocusChain = this.createPath(currFocusedNode);
+
+                    // Find the lowest common ancestor (LCA) between the newly focused node
+                    // and the current focused node.
+                    let lcaIndex = 0;
+                    while (lcaIndex < newFocusChain.length && lcaIndex < currFocusChain.length) {
+                        if (currFocusChain[lcaIndex] !== newFocusChain[lcaIndex]) break;
+                        lcaIndex++;
+                    }
+
+                    // Unset all of the not-common ancestors of the current focused node.
+                    for (let i = lcaIndex; i < currFocusChain.length; i++) {
+                        currFocusChain[i].set(focusedChildString, BrsInvalid.Instance);
+                    }
+                }
+
+                // Set _all_ of the ancestors of the newly focused node.
+                for (let i = 0; i < newFocusChain.length - 1; i++) {
```

Comment:
Consider using `newFocusChain.length - 2` here.  If I'm reading things correctly, `newFocusChain[newFocusChain.length - 1] === this` so the last iteration of this loop would effectively call `this.set(focusedChildString, undefined)`.  Luckily the next line (1484) sets that properly anyway, but it is a little strange.

Reply:
Hmmm I think because i'm using `<` instead of `<=` it executes correctly. Consider:
```
focusChain = [ grandparent, parent, this ]
length = 3 so iterate break once i = 3 - 1 = 2

1. i = 0
    grandparent.set(parent)
2. i = 1
    parent.set(this)
3. i = 2
    break for loop
```